### PR TITLE
sql: ensure that errors in SHOW SYNTAX results get recorded

### LINKED ISSUE #24432: sql: errors from `SHOW SYNTAX` are not counted correctly
Our unit tests use a db/sql connection to send statements and ensure that the error counts and tracking are updated as expected... however, manually recreating that test in a shell, no counts are updated. It looks like the `SHOW SYNTAX` wrapper used by the client is side-stepping the error handling and counting we would otherwise do in the executor.

### PR DESCRIPTION
**Pull request by knz**
Fixes  #24432.
Requested by @awoods187 

SHOW SYNTAX translates parse errors into valid results towards
clients. However we want these errors to be collected for stats
reporting too. This patch changes the logic to ensure that a
translated error gets captured.

Note that only SHOW SYNTAX statements at the top level of a query are
handled in this way (and the errors collected), which is the most
common case and incidentally the case used by `cockroach sql` to do
client-side checking of statements during interactive sessions. If
SHOW SYNTAX is used as a data source (e.g. `SELECT * FROM [SHOW SYNTAX
...]`), the mechanism does not kick in. This is because the
infrastructure necessary for error collection is not easily available
on this code path. This can be extended later if need arises.

Release note (sql change): Errors detected by `SHOW SYNTAX` are now
tracked internally like other SQL errors.

### CODE REVIEW

**Comment by knz** [comment]
@dt any suggestion as to how to test this?

**Comment by dt** [comment]
Maybe run a  `SHOW SYNTAX ...` on something that has a syntax error and on two things that are unimplemented, then look at the executor's error counts. The by-code map i'd expect 1 and 2 for the parse and unimplemented codes, then the unimplemented counts should have the two features with 1 each?

File: pkg/sql/executor.go [hunk 1]
```diff
@@ -1357,7 +1357,9 @@ func runObserverStatement(session *Session, res StatementResult, stmt Statement)
 		if err := runShowSyntax(session.Ctx(), sqlStmt.Statement,
 			func(ctx context.Context, field, msg string) error {
 				return res.AddRow(ctx, tree.Datums{tree.NewDString(field), tree.NewDString(msg)})
-			}); err != nil {
+			},
+			func(err error) {}, /* reportErr */
```
Comment (dt) [inline pkg/sql/executor.go]: huh, why doesn't this one need recording?

**Comment by knz** [comment]
@andreimatei @dt  I have an issue. The `recordError()` method is on the `Server` which is a conn_executor-specific facility. I need to have it available from the `planner` so that I can call it from `ShowSyntax()`. Where is the best place to put the error recording stuff so I see it from `planner`? Is `ExecutorConfig`.

**Comment by knz** [comment]
Is ExecutorConfig a good place? We have that in every planner.

**Comment by andreimatei** [comment]
seems to me that keeping the error reporting on the `Server` and making it an optional attribute of the `planner` is the simplest thing. I think we care about it for statements planned through the `connExecutor` only.

---

Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/sql/executor.go, line 1361 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24819#-LADFoVT1Mp2hFTYUYaj-r1-1361:-LAE7R-R9sNEO3YzfDMK:bkrozsq) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4f8c7c78dbb3c7fd47b5481526caa6498d40460/pkg/sql/executor.go#L1361)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

huh, why doesn't this one need recording?
</blockquote></details>

please don't change this file any more

---

*[pkg/sql/show_syntax.go, line 67 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24819#-LAE7dsD2eP9jz-I3d3s:-LAE7dsD2eP9jz-I3d3t:b-hzwh31) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4f8c7c78dbb3c7fd47b5481526caa6498d40460/pkg/sql/show_syntax.go#L67)):*
> ```Go
> 	stmts, err := parser.Parse(stmt)
> 	if err != nil {
> 		// This error won't propagate to the client as an error, but as a
> ```

nit: move this comment so it's a comment on the `reportErr` argument, and rephrase it to read as a hint as to how the argument is used by the caller

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24819#-:-LAE86ax3QQE1tQlbPty:b-1tsg0h)*

**Comment by knz** [comment]
RFAL

**Comment by knz** [comment]
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/executor.go, line 1361 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24819#-LADFoVT1Mp2hFTYUYaj-r1-1361:-LAFN3ND3BF1jCvmKr76:bb6mt9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4f8c7c78dbb3c7fd47b5481526caa6498d40460/pkg/sql/executor.go#L1361)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please don't change this file any more
</blockquote></details>

:+1:

---

*[pkg/sql/show_syntax.go, line 67 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24819#-LAE7dsD2eP9jz-I3d3s:-LAFN4nmDPXWhobQZnbt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4f8c7c78dbb3c7fd47b5481526caa6498d40460/pkg/sql/show_syntax.go#L67)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: move this comment so it's a comment on the `reportErr` argument, and rephrase it to read as a hint as to how the argument is used by the caller
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24819)*

**Comment by andreimatei** [comment]
:lgtm:

---

Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/sql/show_syntax.go, line 46 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24819#-LAJFAfX6NA44wvatY2F:-LAJFAfY7cjc87KRRVDl:bkapnue) ([raw file](https://github.com/cockroachdb/cockroach/blob/562f806af3d699bce6b319a532c690de1050b29f/pkg/sql/show_syntax.go#L46)):*
> ```Go
> 
> 	comma := ""
> 	// TODO(knz): in the call below, reportErr is nil although we might
> ```

can we just get rid of this method than and not support any wrapping around a `SHOW SYNTAX`? I'd prefer that.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24819#-:-LAJFSDr97kynGDjThgH:bnfp4nl)*

**Comment by knz** [comment]
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/show_syntax.go, line 46 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24819#-LAJFAfX6NA44wvatY2F:-LANq6BJ-edLMDpBZIxc:bx7g3qm) ([raw file](https://github.com/cockroachdb/cockroach/blob/562f806af3d699bce6b319a532c690de1050b29f/pkg/sql/show_syntax.go#L46)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

can we just get rid of this method than and not support any wrapping around a `SHOW SYNTAX`? I'd prefer that.
</blockquote></details>

I don't see any clear reason why we should. In fact, the more statements we can combine with others, the better. The only statements which I really see cannot be combined in that way are BEGIN/SAVEPOINT/etc (txn control).

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24819)*

**Comment by knz** [comment]
Thanks for the reviews!
